// ts-simple-ast/dist/compiler/tools/LanguageService.js
"use strict";
var __values = (this && this.__values)/* istanbul ignore next */ || function (o) {
    var m = typeof Symbol === "function" && o[Symbol.iterator], i = 0;
    if (m) return m.call(o);
    return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
};
var __read = (this && this.__read)/* istanbul ignore next */ || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spread = (this && this.__spread)/* istanbul ignore next */ || function () {
    for (var ar = [], i = 0; i < arguments.length; i++) ar = ar.concat(__read(arguments[i]));
    return ar;
};
Object.defineProperty(exports, "__esModule", { value: true });
var ts = require("typescript");
var manipulation_1 = require("./../../manipulation");
var errors = require("./../../errors");
var utils_1 = require("./../../utils");
var Program_1 = require("./Program");
var results_1 = require("./results");
var LanguageService = /** @class */ (function () {
    /** @internal */
    function LanguageService(global) {
        var _this = this;
        this.sourceFiles = [];
        this.global = global;
        // I don't know what I'm doing for some of this...
        var version = 0;
        var fileExistsSync = function (path) { return _this.global.compilerFactory.containsSourceFileAtPath(path) || global.fileSystem.fileExistsSync(path); };
        var languageServiceHost = {
            getCompilationSettings: function () { return global.compilerOptions; },
            getNewLine: function () { return global.manipulationSettings.getNewLineKind(); },
            getScriptFileNames: function () { return _this.sourceFiles.map(function (s) { return s.getFilePath(); }); },
            getScriptVersion: function (fileName) {
                return (version++).toString();
            },
            getScriptSnapshot: function (fileName) {
                if (!fileExistsSync(fileName))
                    return undefined;
                return ts.ScriptSnapshot.fromString(_this.global.compilerFactory.getSourceFileFromFilePath(fileName).getFullText());
            },
            getCurrentDirectory: function () { return global.fileSystem.getCurrentDirectory(); },
            getDefaultLibFileName: function (options) { return ts.getDefaultLibFilePath(global.compilerOptions); },
            useCaseSensitiveFileNames: function () { return true; },
            readFile: function (path, encoding) {
                if (_this.global.compilerFactory.containsSourceFileAtPath(path))
                    return _this.global.compilerFactory.getSourceFileFromFilePath(path).getFullText();
                return _this.global.fileSystem.readFile(path, encoding);
            },
            fileExists: fileExistsSync,
            directoryExists: function (dirName) { return _this.global.compilerFactory.containsFileInDirectory(dirName) || _this.global.fileSystem.directoryExistsSync(dirName); }
        };
        this.compilerHost = {
            getSourceFile: function (fileName, languageVersion, onError) {
                return _this.global.compilerFactory.getSourceFileFromFilePath(fileName).compilerNode;
            },
            // getSourceFileByPath: (...) => {}, // not providing these will force it to use the file name as the file path
            // getDefaultLibLocation: (...) => {},
            getDefaultLibFileName: function (options) { return languageServiceHost.getDefaultLibFileName(options); },
            writeFile: function (filePath, data, writeByteOrderMark, onError, sourceFiles) {
                utils_1.FileUtils.ensureDirectoryExistsSync(_this.global.fileSystem, utils_1.FileUtils.getDirPath(filePath));
                _this.global.fileSystem.writeFileSync(filePath, data);
            },
            getCurrentDirectory: function () { return languageServiceHost.getCurrentDirectory(); },
            getDirectories: function (path) {
                console.log("ATTEMPT TO GET DIRECTORIES");
                return [];
            },
            fileExists: function (fileName) { return languageServiceHost.fileExists(fileName); },
            readFile: function (fileName) { return languageServiceHost.readFile(fileName); },
            getCanonicalFileName: function (fileName) { return utils_1.FileUtils.getStandardizedAbsolutePath(fileName); },
            useCaseSensitiveFileNames: function () { return languageServiceHost.useCaseSensitiveFileNames(); },
            getNewLine: function () { return languageServiceHost.getNewLine(); },
            getEnvironmentVariable: function (name) { return process.env[name]; }
        };
        this._compilerObject = ts.createLanguageService(languageServiceHost);
    }
    Object.defineProperty(LanguageService.prototype, "compilerObject", {
        /**
         * Gets the compiler language service.
         */
        get: function () {
            return this._compilerObject;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Resets the program. This should be done whenever any modifications happen.
     * @internal
     */
    LanguageService.prototype.resetProgram = function () {
        if (this.program != null)
            this.program.reset(this.sourceFiles.map(function (s) { return s.getFilePath(); }), this.compilerHost);
    };
    /**
     * Gets the language service's program.
     */
    LanguageService.prototype.getProgram = function () {
        if (this.program == null)
            this.program = new Program_1.Program(this.global, this.sourceFiles.map(function (s) { return s.getFilePath(); }), this.compilerHost);
        return this.program;
    };
    /**
     * Rename the specified node.
     * @param node - Node to rename.
     * @param newName - New name for the node.
     */
    LanguageService.prototype.renameNode = function (node, newName) {
        errors.throwIfNotStringOrWhitespace(newName, "newName");
        if (node.getText() === newName)
            return;
        this.renameLocations(this.findRenameLocations(node), newName);
    };
    /**
     * Rename the provided rename locations.
     * @param renameLocations - Rename locations.
     * @param newName - New name for the node.
     */
    LanguageService.prototype.renameLocations = function (renameLocations, newName) {
        var renameLocationsBySourceFile = new utils_1.KeyValueCache();
        try {
            for (var renameLocations_1 = __values(renameLocations), renameLocations_1_1 = renameLocations_1.next(); !renameLocations_1_1.done; renameLocations_1_1 = renameLocations_1.next()) {
                var renameLocation = renameLocations_1_1.value;
                var locations = renameLocationsBySourceFile.getOrCreate(renameLocation.getSourceFile(), function () { return []; });
                locations.push(renameLocation);
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (renameLocations_1_1 && !renameLocations_1_1.done && (_a = renameLocations_1.return)) _a.call(renameLocations_1);
            }
            finally { if (e_1) throw e_1.error; }
        }
        try {
            for (var _b = __values(renameLocationsBySourceFile.getEntries()), _c = _b.next(); !_c.done; _c = _b.next()) {
                var _d = __read(_c.value, 2), sourceFile = _d[0], locations = _d[1];
                var difference = 0;
                try {
                    for (var _e = __values(locations.map(function (l) { return l.getTextSpan(); })), _f = _e.next(); !_f.done; _f = _e.next()) {
                        var textSpan = _f.value;
                        var start = textSpan.getStart();
                        start -= difference;
                        manipulation_1.replaceNodeText(sourceFile, start, start + textSpan.getLength(), newName);
                        difference += textSpan.getLength() - newName.length;
                    }
                }
                catch (e_2_1) { e_2 = { error: e_2_1 }; }
                finally {
                    try {
                        if (_f && !_f.done && (_g = _e.return)) _g.call(_e);
                    }
                    finally { if (e_2) throw e_2.error; }
                }
            }
        }
        catch (e_3_1) { e_3 = { error: e_3_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_h = _b.return)) _h.call(_b);
            }
            finally { if (e_3) throw e_3.error; }
        }
        var e_1, _a, e_3, _h, e_2, _g;
    };
    /**
     * Gets the definitions for the specified node.
     * @param node - Node.
     */
    LanguageService.prototype.getDefinitions = function (node) {
        return this.getDefinitionsAtPosition(node.sourceFile, node.getStart());
    };
    /**
     * Gets the definitions at the specified position.
     * @param sourceFile - Source file.
     * @param pos - Position.
     */
    LanguageService.prototype.getDefinitionsAtPosition = function (sourceFile, pos) {
        var _this = this;
        var results = this.compilerObject.getDefinitionAtPosition(sourceFile.getFilePath(), pos) || [];
        return results.map(function (info) { return new results_1.DefinitionInfo(_this.global, info); });
    };
    /**
     * Gets the implementations for the specified node.
     * @param node - Node.
     */
    LanguageService.prototype.getImplementations = function (node) {
        return this.getImplementationsAtPosition(node.sourceFile, node.getStart());
    };
    /**
     * Gets the implementations at the specified position.
     * @param sourceFile - Source file.
     * @param pos - Position.
     */
    LanguageService.prototype.getImplementationsAtPosition = function (sourceFile, pos) {
        var _this = this;
        var results = this.compilerObject.getImplementationAtPosition(sourceFile.getFilePath(), pos) || [];
        return results.map(function (location) { return new results_1.ImplementationLocation(_this.global, location); });
    };
    /**
     * Finds references based on the specified node.
     * @param node - Node to find references for.
     */
    LanguageService.prototype.findReferences = function (node) {
        return this.findReferencesAtPosition(node.sourceFile, node.getStart());
    };
    /**
     * Finds references based on the specified position.
     * @param sourceFile - Source file.
     * @param pos - Position to find the reference at.
     */
    LanguageService.prototype.findReferencesAtPosition = function (sourceFile, pos) {
        var _this = this;
        var results = this.compilerObject.findReferences(sourceFile.getFilePath(), pos) || [];
        return results.map(function (s) { return new results_1.ReferencedSymbol(_this.global, s); });
    };
    /**
     * Find the rename locations for the specified node.
     * @param node - Node to get the rename locations for.
     */
    LanguageService.prototype.findRenameLocations = function (node) {
        var _this = this;
        var sourceFile = node.getSourceFile();
        var renameLocations = this.compilerObject.findRenameLocations(sourceFile.getFilePath(), node.getStart(), false, false) || [];
        return renameLocations.map(function (l) { return new results_1.RenameLocation(_this.global, l); });
    };
    /**
     * @internal
     */
    LanguageService.prototype.addSourceFile = function (sourceFile) {
        // todo: these source files should be strictly stored in the factory cache
        this.sourceFiles.push(sourceFile);
        this.resetProgram();
    };
    /**
     * @internal
     */
    LanguageService.prototype.removeSourceFile = function (sourceFile) {
        var index = this.sourceFiles.indexOf(sourceFile);
        if (index === -1)
            return false;
        this.sourceFiles.splice(index, 1);
        this.resetProgram();
        sourceFile.dispose(); // todo: don't dispose, just remove the language service for this node
        return true;
    };
    /**
     * @internal
     */
    LanguageService.prototype.getSourceFiles = function () {
        return __spread(this.sourceFiles);
    };
    return LanguageService;
}());
exports.LanguageService = LanguageService;
