// ts-simple-ast/dist/manipulation/insertion/insertIntoCommaSeparatedNodes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ts = require("typescript");
var insertIntoCreatableSyntaxList_1 = require("./insertIntoCreatableSyntaxList");
function insertIntoCommaSeparatedNodes(opts) {
    var currentNodes = opts.currentNodes, insertIndex = opts.insertIndex, newTexts = opts.newTexts, parent = opts.parent;
    var nextNode = currentNodes[insertIndex];
    var previousNode = currentNodes[insertIndex - 1];
    var numberOfSyntaxListItemsInserting = newTexts.length * 2 - 1;
    if (nextNode != null) {
        insertIntoCreatableSyntaxList_1.insertIntoCreatableSyntaxList({
            parent: parent,
            insertPos: nextNode.getStart(),
            newText: newTexts.join(", ") + ", ",
            syntaxList: nextNode.getParentSyntaxListOrThrow(),
            childIndex: nextNode.getChildIndex(),
            insertItemsCount: numberOfSyntaxListItemsInserting + 1 // extra comma
        });
    }
    else if (previousNode != null) {
        insertIntoCreatableSyntaxList_1.insertIntoCreatableSyntaxList({
            parent: parent,
            insertPos: previousNode.getEnd(),
            newText: ", " + newTexts.join(", "),
            syntaxList: previousNode.getParentSyntaxListOrThrow(),
            childIndex: previousNode.getChildIndex() + 1,
            insertItemsCount: numberOfSyntaxListItemsInserting + 1 // extra comma
        });
    }
    else {
        insertIntoCreatableSyntaxList_1.insertIntoCreatableSyntaxList({
            parent: parent,
            insertPos: parent.getFirstChildByKindOrThrow(ts.SyntaxKind.SyntaxList).getPos(),
            syntaxList: parent.getFirstChildByKind(ts.SyntaxKind.SyntaxList),
            newText: newTexts.join(", "),
            childIndex: 0,
            insertItemsCount: numberOfSyntaxListItemsInserting
        });
    }
}
exports.insertIntoCommaSeparatedNodes = insertIntoCommaSeparatedNodes;
