// code-block-writer/dist/code-block-writer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var CodeBlockWriter = (function () {
    function CodeBlockWriter(opts) {
        this._currentIndentation = 0;
        this._text = "";
        this._isAtStartOfBlock = false;
        this._newLine = (opts && opts.newLine) || "\n";
        this._indentationText = getIndentationText((opts && opts.useTabs) || false, (opts && opts.indentNumberOfSpaces) || 4);
    }
    CodeBlockWriter.prototype.block = function (block) {
        this.spaceIfLastNotSpace();
        this.inlineBlock(block);
        this.newLine();
        return this;
    };
    CodeBlockWriter.prototype.inlineBlock = function (block) {
        this.write("{");
        this._currentIndentation++;
        this.newLine();
        this._isAtStartOfBlock = true;
        block();
        this.removeLastIfNewLine();
        this._currentIndentation--;
        this.newLineIfLastNotNewLine().write("}");
        return this;
    };
    CodeBlockWriter.prototype.conditionalWriteLine = function (condition, str) {
        if (condition)
            this.writeLine(str);
        return this;
    };
    CodeBlockWriter.prototype.writeLine = function (str) {
        this.newLineIfLastNotNewLine();
        this.writeIndentingNewLines(str);
        this.newLine();
        return this;
    };
    CodeBlockWriter.prototype.newLineIfLastNotNewLine = function () {
        if (!this.isLastCharANewLine())
            this.newLine();
        return this;
    };
    CodeBlockWriter.prototype.blankLine = function () {
        return this.newLine().newLine();
    };
    CodeBlockWriter.prototype.indent = function () {
        return this.write(this._indentationText);
    };
    CodeBlockWriter.prototype.conditionalNewLine = function (condition) {
        if (condition)
            this.newLine();
        return this;
    };
    CodeBlockWriter.prototype.newLine = function () {
        var willCreateAConsecutiveBlankLine = this.isLastLineBlankLine() && this.isCurrentLineBlank();
        if (!willCreateAConsecutiveBlankLine && !this._isAtStartOfBlock && this._text.length !== 0)
            this.baseWrite(this._newLine);
        return this;
    };
    CodeBlockWriter.prototype.spaceIfLastNotSpace = function () {
        var lastChar = this.getLastChar();
        if (lastChar != null && lastChar !== " " && !this.isLastCharANewLine())
            this.baseWrite(" ");
        return this;
    };
    CodeBlockWriter.prototype.conditionalWrite = function (condition, str) {
        if (condition)
            this.write(str);
        return this;
    };
    CodeBlockWriter.prototype.write = function (str) {
        this.writeIndentingNewLines(str);
        return this;
    };
    CodeBlockWriter.prototype.getLength = function () {
        return this._text.length;
    };
    CodeBlockWriter.prototype.toString = function () {
        return this.removeConsecutiveNewLineAtEndOfString(this._text);
    };
    CodeBlockWriter.prototype.writeIndentingNewLines = function (str) {
        var _this = this;
        var items = (str || "").split(/\r?\n/);
        items.forEach(function (s, i) {
            if (i > 0)
                _this.baseWrite(_this._newLine);
            _this.baseWrite(s);
            if (i > 0 && i === items.length - 1 && s.length === 0)
                _this.baseWrite(_this._newLine);
        });
    };
    CodeBlockWriter.prototype.baseWrite = function (str) {
        this._isAtStartOfBlock = false;
        if (str == null || str.length === 0)
            return this;
        if (str !== this._newLine && this.isLastCharANewLine())
            this.writeIndentation();
        this._text += str;
        return this;
    };
    CodeBlockWriter.prototype.removeConsecutiveNewLineAtEndOfString = function (text) {
        var consecutiveNewline = this._newLine + this._newLine;
        var lastIndexOfConsecutiveNewLines = text.lastIndexOf(consecutiveNewline);
        if (lastIndexOfConsecutiveNewLines >= 0 && lastIndexOfConsecutiveNewLines === text.length - consecutiveNewline.length)
            text = text.substr(0, text.length - this._newLine.length);
        return text;
    };
    CodeBlockWriter.prototype.removeLastIfNewLine = function () {
        if (this.isLastLineBlankLine() && this.isCurrentLineBlank())
            this._text = this._text.substr(0, this._text.length - this._newLine.length);
    };
    CodeBlockWriter.prototype.isCurrentLineBlank = function () {
        return this.getCurrentLine().length === 0;
    };
    CodeBlockWriter.prototype.isLastLineBlankLine = function () {
        return this.getLastLine() === this._newLine;
    };
    CodeBlockWriter.prototype.getCurrentLine = function () {
        var lastNewLineIndex = this._text.lastIndexOf(this._newLine);
        if (lastNewLineIndex >= 0)
            return this._text.substr(lastNewLineIndex + this._newLine.length);
        else
            return "";
    };
    CodeBlockWriter.prototype.getLastLine = function () {
        var lastNewLineIndex = this._text.lastIndexOf(this._newLine);
        if (lastNewLineIndex < 0)
            return null;
        var secondLastNewLineIndex = this._text.lastIndexOf(this._newLine, lastNewLineIndex - 1);
        if (secondLastNewLineIndex === -1)
            secondLastNewLineIndex = 0;
        return this._text.substr(secondLastNewLineIndex, lastNewLineIndex - secondLastNewLineIndex);
    };
    CodeBlockWriter.prototype.isLastCharANewLine = function () {
        return this._text.indexOf(this._newLine, this._text.length - this._newLine.length) !== -1;
    };
    CodeBlockWriter.prototype.getLastChar = function () {
        var lastChar = null;
        if (this._text.length > 0)
            lastChar = this._text[this._text.length - 1];
        return lastChar;
    };
    CodeBlockWriter.prototype.writeIndentation = function () {
        this._text += Array(this._currentIndentation + 1).join(this._indentationText);
    };
    return CodeBlockWriter;
}());
exports.default = CodeBlockWriter;
function getIndentationText(useTabs, numberSpaces) {
    if (useTabs)
        return "\t";
    return Array(numberSpaces + 1).join(" ");
}
